frontend/controller/actions/chatroom.js

```diff
@@ -3,13 +3,52 @@ import sbp from '@sbp/sbp'
 
 import { GIErrorUIRuntimeError, L } from '@common/common.js'
 import { has, omit } from '@model/contracts/shared/giLodash.js'
+import { CHATROOM_TYPES, MESSAGE_RECEIVE_RAW, MESSAGE_TYPES } from '@model/contracts/shared/constants.js'
 import { GIMessage } from '~/shared/domains/chelonia/GIMessage.js'
 import { Secret } from '~/shared/domains/chelonia/Secret.js'
 import { encryptedOutgoingData, encryptedOutgoingDataWithRawKey } from '~/shared/domains/chelonia/encryptedData.js'
 // Using relative path to crypto.js instead of ~-path to workaround some esbuild bug
 import { CURVE25519XSALSA20POLY1305, EDWARDS25519SHA512BATCH, deserializeKey, keyId, keygen, serializeKey } from '../../../shared/domains/chelonia/crypto.js'
 import type { GIRegParams } from './types.js'
 import { encryptedAction, encryptedNotification } from './utils.js'
+import { makeMentionFromUserID } from '@model/chatroom/utils.js'
+import messageReceivePostEffect from '@model/notifications/messageReceivePostEffect.js'
+
+sbp('okTurtles.events/on', MESSAGE_RECEIVE_RAW, ({
+  contractID,
+  data,
+  innerSigningContractID,
+  newMessage
```

Comment:
Please add a comment here to explain that if `newMessage` is undefined, it means an existing message is being edited.